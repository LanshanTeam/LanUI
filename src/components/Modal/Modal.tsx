import { createContext, ReactNode } from 'react';
import { createPortal } from 'react-dom';
import ModalOverlay from './components/ModalOverlay';
import ModalContent from './components/ModalContent';
import ModalHeader from './components/ModalHeader';
import ModalBody from './components/ModalBody';
import ModalFooter from './components/ModalFooter';
import './style/modal.less';

export interface ModalProps {
    /**
     * @desc 是否打开
     */
    isOpen: Boolean;
    /**
     * @desc 打开回调函数
     */
    onOk?: Function;
    /**
     * @desc 关闭回调函数
     */
    onCancel?: Function;
    /**
     * @desc 宽度
     */
    width?: string;
    /**
     * @desc 标题
     */
    title?: string;
    /**
     * @desc 页脚内容
     */
    footer?: ReactNode;
    /**
     * @desc 用户传递的样式
     */
    style?: Object;
    /**
     * @desc 主体内容
     */
    children?: ReactNode;
}

export interface ModalContext extends Partial<ModalProps> {
    onOk?: Function;
    onCancel?: Function;
}

export const ModalContext = createContext<ModalContext>({
    isOpen: undefined,
    onOk: undefined,
    onCancel: undefined,
});

const Modal = ({
    isOpen,
    onOk,
    onCancel,
    width = '520px',
    title = 'title',
    footer,
    style,
    children,
}: ModalProps) => {
    return (
        isOpen && (
            <ModalContext.Provider value={{ isOpen, onOk, onCancel }}>
                {createPortal(
                    <>
                        <ModalOverlay onClick={() => onCancel!()} />
                        <ModalContent styles={{ width, ...style }}>
                            <ModalHeader title={title}></ModalHeader>
                            <ModalBody>{children}</ModalBody>
                            <ModalFooter>{footer}</ModalFooter>
                        </ModalContent>
                    </>,
                    document.body
                )}
            </ModalContext.Provider>
        )
    );
};

export default Modal;
