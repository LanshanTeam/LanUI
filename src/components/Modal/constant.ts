/**
 * @desc 进场出场动画持续时间
 */
export const DURATION = 0.1;
