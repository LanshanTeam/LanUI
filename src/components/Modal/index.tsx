import Modal from './Modal';
import '../../style/index';
export default Modal;
