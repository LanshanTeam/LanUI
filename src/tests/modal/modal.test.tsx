import React from 'react';
import { render, fireEvent, screen, act } from '@testing-library/react';
import Modal from '../../components/Modal/index';

jest.useFakeTimers();

describe('Modal Component', () => {
    it('should renders closed Modal by default', () => {
        render(
            <Modal isOpen={false} onOk={() => {}} onCancel={() => {}}>
                Test Content
            </Modal>
        );
        expect(screen.queryByText('Test Content')).toBeNull();
    });

    it('should render correctly', () => {
        render(
            <Modal isOpen={true} onOk={() => {}} onCancel={() => {}}>
                Test Content
            </Modal>
        );
        expect(screen.queryByText('Test Content')).toMatchSnapshot();
    });

    it('should renders Modal with custom footer', () => {
        render(
            <Modal
                isOpen={true}
                onOk={() => {}}
                onCancel={() => {}}
                footer={<button onClick={() => {}}>自定义确认</button>}
            >
                Test Content
            </Modal>
        );

        expect(screen.queryByText('Test Content')).toBeInTheDocument();
    });

    it('calls onOk callback when Ok button is clicked', () => {
        const onOkMock = jest.fn();
        render(
            <Modal isOpen={true} onOk={onOkMock} onCancel={() => {}}>
                Test Content
            </Modal>
        );
        fireEvent.click(screen.getByText('OK'));
        act(() => {
            jest.runAllTimers();
        });
        expect(onOkMock).toHaveBeenCalled();
    });

    it('calls onCancel callback when Cancel button is clicked', () => {
        const onCancelMock = jest.fn();
        render(
            <Modal isOpen={true} onOk={() => {}} onCancel={onCancelMock}>
                Test Content
            </Modal>
        );
        fireEvent.click(screen.getByText('Cancel'));
        act(() => {
            jest.runAllTimers();
        });
        expect(onCancelMock).toHaveBeenCalled();
    });
});
