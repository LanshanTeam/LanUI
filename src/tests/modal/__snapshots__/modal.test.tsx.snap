// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component should render correctly 1`] = `
<div
  class="modal-content"
  style="width: 520px; opacity: 1; top: 50%; transition: all 0.1s;"
>
  <span
    class="i_modal_close"
  >
    X
  </span>
  <h2>
    title
  </h2>
  Test Content
  <div
    class="btn-area"
  >
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      type="button"
    >
      Cancel
    </button>
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      style="margin-left: 0.625rem;"
      type="button"
    >
      OK
    </button>
  </div>
</div>
`;
